### ISSUE TITLE
IX-F JSON postmortem/preview for networks

### ISSUE THREAD

**Issue by arnoldnipper** [peeringdb/peeringdb#505]
With release 2.13.0 IXes have full insight what the IX-F JSON importer will do. However network entries still may go away w/o any notification.

The idea is to provide two features
* postmortem
  * analyse up to $configurable imports for changes for $network
* preview
  * analyse what would happen to $network with the next import

**Comment by arnoldnipper** [comment]
may relate tp #499

**Comment by tardoe** [comment]
Upvote for this one - when pushing an IX-F feed, the network should have full visibility into the changes it'll cause. This will limit the blast radius for new feature deployment and custom-backend stack refactoring (aka testing). This will also recreate the time to resolution for ticket involving incorrect data export - e.g. allows the network to make a change and instantly validate it'll ultimately fix the issue.

**Comment by robwdwd** [comment]
Why not use a mismatch flag? Where any mismatch between the IX data and the network operator data is simply flagged. No data should be changed or removed from the network operator record. 

The flag can trigger an email notification an alert in the web front end. The network operator can act on that as they see fit. If it's caused by an IX data error they can ignore it until it's fixed (or contact the IX) or correct a mistake (in the ip address etc) made by the network operator.

Flag can even be in data from the API so any other network operators can see that this IX connection entry could potentially have an error and can act accordingly when configuring a new peer neighbour.

Basically silently removing entries is very bad.

**Comment by Shawna** [comment]
I understand **preview** as essentially the same UI as for the exchange implementation, just filtered to a specific network, is that correct?

Also where do you want to tie these into the UI, i can see the **preview** as a button next to the exchange name in "Public Peering Exchange Points" on the network view - although it'd have to be a tiny button since space is already an issue. 

For **postmortem** (since the entry may be deleted) maybe have a separate UI (Run import postmortem for IX) that lets them select an exchange. Maybe that would work for preview as well. Please advise.

**Comment by arnoldnipper** [comment]
> I understand **preview** as essentially the same UI as for the exchange implementation, just filtered to a specific network, is that correct?

Right, **preview** for the network should have essentially the same UI (i.e. GUI as well as API) as for the exchange implementation. However, for a given network it has to show all actions for all IXes that provide IX-F JSON import.

Hence imho a button in the `PeeringDB Configuration` section would make more sense. The **postmortem** button then could go there as well.

**Comment by arnoldnipper** [comment]
@robwdwd are you Ok when also taking #518 into account?

**Comment by ccaputo** [comment]
Related: https://github.com/peeringdb/peeringdb/issues/540

**Comment by vegu** [comment]
@arnoldnipper since the network version of this will need to show the ix name (and link to the ix) i propose we add `ix_name` and `ix_id` fields to the `peer` object of the ix-f import log schema - any reservations?

## Example

```
        "net_id": 20, 
        "speed": 10000, 
        "ipaddr4": "206.41.110.51", 
        "ipaddr6": "2001:504:41:110::51", 
        "is_rs_peer": false, 
        "ixlan_id": 239, 
        "asn": 63311
```

becomes

```
        "ix_id": 239, 
        "ix_name": "ChIX", 
        "net_id": 20, 
        "speed": 10000, 
        "ipaddr4": "206.41.110.51", 
        "ipaddr6": "2001:504:41:110::51", 
        "is_rs_peer": false, 
        "ixlan_id": 239, 
        "asn": 63311
```

**Comment by arnoldnipper** [comment]
> ince the network version of this will need to show the ix name (and link to the ix) i propose we add `ix_name` and `ix_id` fields to the `peer` object of the ix-f import log schema - any reservations?

Even better with `ix_id` and `ix_name`, though it's adding redundancy. Or to be more precise. That will be shown when you retrieve the record. But you will not be able to **set** it. Right? I.e. read-only.

Would make sense for the `ixpfx` record as well while you already are it. Agree?

**Comment by vegu** [comment]
@arnoldnipper the change i meant would only be in the IX-F import preview log output (as in https://www.peeringdb.com/import/ixlan/{id}/ixf/preview) not the api. They dont share serializers so we can make changes to one without making changes to the other - if these fields make sense to have in the api as well that should be a separate ticket.

**Comment by arnoldnipper** [comment]
Agree, @vegu ... imho it could make sense to get as complete information (esp. ids) on return just to save additional calls to the API

**Comment by Shawna** [comment]
#505

The Postmortem views on beta will look a bit empty until import actually
has done some changes. This is because old changes never logged the
reasoning for the change, so we are not showing those, since users would
not be able to understand why a change happened.

The preview functionality works the same way as the one for exchanges -
however since it needs to do it for many exchanges at once it is working
off of cached IX-F data. This data is updated every time the importer runs.

If we need a higher frequency of updating those caches we can run a
command to just do that without doing any importing.

Both tools can be found in the PeeringDB Config section of the network
view as long as the viewing user has write permissions to the network.

Both tools are expensive and are rate limited to stop any potential abuse.
